### PR TITLE
Run typecheck and eslint on `./run lint`

### LINKED ISSUE #6244: Integrate ESLint with `./run lint`
The ESLint is not integrated with our `./run lint` command. Because of that it is not enforced upon PRs and prone to regressions.

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
Fixes #6244

### Important Notes
~~The lint check *will fail* because there are eslint errors. #6267 should fix this.~~
The typecheck check should be tested by making some code invalid. the simplest way to do this would be to change a variable name.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by somebody1234** [comment]
not sure whether i'm missing something (since a 3 line change is a *really* small change) but it seems to work
styles *should* be fine too since `./run lint` also runs and diffs against rust formatter output.

<details>
<summary>Output from new checks</summary>
<pre><code>TRACE ide_ci::program::command: npmℹ️ <START>
TRACE ide_ci::program::command: npm⚠️ <START>
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️ > root@1.0.0 typecheck
 INFO ide_ci::program::command: npmℹ️ > npm run typecheck --workspace enso; npm run typecheck --workspace enso-content; npm run typecheck --workspace enso-dashboard; npm run typecheck --workspace enso-authentication
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️ > enso@0.0.0-dev typecheck
 INFO ide_ci::program::command: npmℹ️ > tsc --noEmit
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️ > enso-content@1.0.0 typecheck
 INFO ide_ci::program::command: npmℹ️ > tsc --noEmit
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️ > enso-dashboard@0.1.0 typecheck
 INFO ide_ci::program::command: npmℹ️ > tsc --noEmit
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️
 INFO ide_ci::program::command: npmℹ️ > enso-authentication@1.0.0 typecheck
 INFO ide_ci::program::command: npmℹ️ > tsc --noEmit
 INFO ide_ci::program::command: npmℹ️
TRACE ide_ci::program::command: npmℹ️ <ENDUT>
TRACE ide_ci::program::command: npm⚠️ <ENDUT>
 INFO main_internal:Running process.{command=Command:
	cd "/home/me/git/enso/app/ide-desktop" && "/usr/bin/npm" "run" "typecheck"
	with working directory: /home/me/git/enso/app/ide-desktop pid=27315 status=0}: ide_ci::program::command: close
 INFO main_internal:Running process.: ide_ci::program::command: new
DEBUG main_internal:Running process.{command=Command:
	cd "/home/me/git/enso/app/ide-desktop" && "/usr/bin/npx" "eslint" "."
	with working directory: /home/me/git/enso/app/ide-desktop}: ide_ci::program::command: Spawning.
TRACE ide_ci::program::command: npxℹ️ <START>
TRACE ide_ci::program::command: npx⚠️ <START>
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/client/electron-builder-config.ts
 INFO ide_ci::program::command: npxℹ️   22:1  error  Expected 'multiple' syntax before 'single' syntax  sort-imports
 INFO ide_ci::program::command: npxℹ️   22:1  error  No {} imports and exports                          no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/client/src/authentication.ts
 INFO ide_ci::program::command: npxℹ️    77:1   error    Expected 'all' syntax before 'single' syntax                                                          sort-imports
 INFO ide_ci::program::command: npxℹ️    77:1   error    Import node modules before npm modules, our modules, and relative imports, separated by a blank line  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️    78:1   error    Import node modules before npm modules, our modules, and relative imports, separated by a blank line  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️    79:1   error    Import node modules before npm modules, our modules, and relative imports, separated by a blank line  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️   168:21  warning  'err' is already declared in the upper scope on line 161 column 64                                    @typescript-eslint/no-shadow
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/client/src/bin/project-manager.ts
 INFO ide_ci::program::command: npxℹ️   26:9  error  No early returns  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/client/src/file-associations.ts
 INFO ide_ci::program::command: npxℹ️   149:9  error  No early returns  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/client/src/index.ts
 INFO ide_ci::program::command: npxℹ️   285:59  warning  '_event' is defined but never used                                                                                                                                  @typescript-eslint/no-unused-vars
 INFO ide_ci::program::command: npxℹ️   287:17  error    Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/client/src/project-management.ts
 INFO ide_ci::program::command: npxℹ️    53:13  error  No early returns  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️   101:13  error  No early returns  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/client/tasks/signArchivesMacOs.ts
 INFO ide_ci::program::command: npxℹ️   206:13  error  No early returns  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/content/bundle.ts
 INFO ide_ci::program::command: npxℹ️   10:5  error  No early returns  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx
 INFO ide_ci::program::command: npxℹ️   177:19  warning  '_extension' is defined but never used                  @typescript-eslint/no-unused-vars
 INFO ide_ci::program::command: npxℹ️   216:28  warning  'setSelectedAssets' is assigned a value but never used  @typescript-eslint/no-unused-vars
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/ide.tsx
 INFO ide_ci::program::command: npxℹ️   78:32  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion
 INFO ide_ci::program::command: npxℹ️   79:33  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ /home/me/git/enso/app/ide-desktop/lib/icons/src/index.js
 INFO ide_ci::program::command: npxℹ️   217:9  error  No early returns  no-restricted-syntax
 INFO ide_ci::program::command: npxℹ️
 INFO ide_ci::program::command: npxℹ️ ✖ 20 problems (14 errors, 6 warnings)
 INFO ide_ci::program::command: npxℹ️
TRACE ide_ci::program::command: npxℹ️ <ENDUT>
TRACE ide_ci::program::command: npx⚠️ <ENDUT>
 INFO main_internal:Running process.{command=Command:
	cd "/home/me/git/enso/app/ide-desktop" && "/usr/bin/npx" "eslint" "."
	with working directory: /home/me/git/enso/app/ide-desktop pid=27707 status=1}: ide_ci::program::command: close
ERROR main_internal: enso_build_cli: error=Command failed: Command:
	cd "/home/me/git/enso/app/ide-desktop" && "/usr/bin/npx" "eslint" "."
	with working directory: /home/me/git/enso/app/ide-desktop
</code></pre>
</details>

**Review by mwu-tow — APPROVED** [review]
> not sure whether i'm missing something (since a 3 line change is a really small change) but it seems to work

Well, it does not need to be a big change. If it works (please check, if it runs correctly on a clean working copy), then it is fine. :)

**Comment by somebody1234** [comment]
seems to run correctly on ci (fails as expected). i think i'll merge the eslint fixes and verify that it's working, then add a type error to verify that typecheck lint works

**Comment by somebody1234** [comment]
errors on eslint error: :heavy_check_mark: 
errors on typescript error: :heavy_check_mark: (fixed)
succeeds when the eslint fixes are merged in: :heavy_check_mark:

**Comment by PabloBuchu** [comment]
@somebody1234 so QA passed 🟢 how about these lint errors for linux https://github.com/enso-org/enso/actions/runs/4739322557/jobs/8414041628?pr=6314 we can fix them now?

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Run typecheck and eslint on `./run lint` · enso-org/enso@c6b4d57</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by somebody1234** [comment]
@PabloBuchu as mentioned in the original comment, #6267 will fix the lint errors. not sure if lint failing blocks merge

**Comment by mwu-tow** [comment]
@somebody1234 
> @PabloBuchu as mentioned in the original comment, #6267 will fix the lint errors. not sure if lint failing blocks merge

It blocks the merge; this is the main point of integrating this with `./run lint`. :)

**Comment by somebody1234** [comment]
ah, of course... was on mobile at the time so i couldn't tell whether it was required - but the other PR should be just about ready, so i think it *should* be fine to merge the other one, then this one - when the other is confirmed to be ready, of course

**Comment by mwu-tow** [comment]
> ah, of course... was on mobile at the time so i couldn't tell whether it was required - but the other PR should be just about ready, so i think it should be fine to merge the other one, then this one - when the other is confirmed to be ready, of course

Makes total sense!

File: build/cli/src/lib.rs [hunk 1]
```diff
@@ -835,6 +837,12 @@ pub async fn main_internal(config: Option<enso_build::config::Config>) -> Result
                 .await?;
 
             prettier::check(&ctx.repo_root).await?;
+
+            let js_modules_root = ctx.repo_root.join("app/ide-desktop");
+
+            Npm.cmd()?.current_dir(&js_modules_root).args(["run", "typecheck"]).run_ok().await?;
+
+            Npx.cmd()?.current_dir(&js_modules_root).args(["eslint", "."]).run_ok().await?;
```
Comment (wdanilo) [inline build/cli/src/lib.rs]: Why are we creating these newlines here? I don't think they are needed - they just make the code longer.

**Review by wdanilo — APPROVED** [review]

File: build/cli/src/lib.rs (hunk 1 above)
Comment (somebody1234) [reply build/cli/src/lib.rs]: should be fixed now

**Comment by somebody1234** [comment]
confirmed to exit with success because the other PR has been merged in. there are 4 eslint warnings (not errors) for `lib/dashboard`, but this is intentional as dashboard implementation is still in progress.

(two of the warnings should be fixed by another PR, the other two are for features that are not implemented yet)

**Comment by wdanilo** [comment]
If this PR is waiting for other PRs, let's change it to draft. If it is ready to merge - lets merge it :)

**Comment by somebody1234** [comment]
@wdanilo some CIs failing, not sure why + waiting for approval from a code owner

**Comment by somebody1234** [comment]
lint check *is* successful though so it should be safe enough to merge

**Comment by somebody1234** [comment]
@PabloBuchu @indiv0 would one of you mind reviewing? a small QA would be nice but i don't think it's strictly necessary as lint CI is only run on linux.

i've merged develop into this branch *locally* to test (will not *actually* merge in, just checking there will no issues after merging into develop) and will mark as ready for merge if it runs correctly

**Comment by somebody1234** [comment]
oh, actually not sure why some js fixes snuck into this PR, but they're already here anyway.

anyway, lint run has succeeded locally

**Comment by somebody1234** [comment]
note: the "Engine CI / Engine (linux)" is required but failing with output:
i'm going out on a limb and assuming this PR isn't the cause :stuck_out_tongue: 

```
❌ should not duplicate warnings
	An unexpected panic was thrown: (No_Such_Method.Error (SQLite_Connection.Value (Connection.Value (JDBC_Connection.Value org.enso.interpreter.runtime.data.ManagedResource@6b437b24) SQLite_Dialect)) UnresolvedSymbol<upload_table>)
```

**Comment by mwu-tow** [comment]
@somebody1234 

> note: the "Engine CI / Engine (linux)" is required but failing with output:
i'm going out on a limb and assuming this PR isn't the cause 😛

Looks like some random issue, perhaps yet another mixup between different Engine versions across the branches. Pinging @hubertp @radeusgd

**Comment by wdanilo** [comment]
As we need this PR and the fails are engine-related only, merging it now.

**Pull request merged**